File: tools/behavioral-twin/predict.py
#!/usr/bin/env python3
"""
Behavioral Twin predictor for Chief state/timing/tone.

Output JSON contract:
{
  "state": "awake|busy|available|winding-down|sleeping",
  "confidence": 0.0-1.0,
  "recommended_tone": "...",
  "alert_ok": true|false,
  "next_available_window": "YYYY-MM-DDTHH:MM:SS-05:00/YYYY-MM-DDTHH:MM:SS-05:00"
}
"""

from __future__ import annotations

import argparse
import json
import os
import statistics
import subprocess
from dataclasses import dataclass
from datetime import datetime, time, timedelta
from zoneinfo import ZoneInfo

DB_NAME = os.environ.get("CORTANA_DB", "cortana")
ET = ZoneInfo("America/New_York")
PSQL_PATH = "/opt/homebrew/opt/postgresql@17/bin"


@dataclass
class PatternSummary:
    wake_minutes_by_dow: dict[int, list[int]]
    sleep_minutes_by_dow: dict[int, list[int]]
    observed_hours_by_dow: dict[int, dict[int, int]]
    sample_count: int


def _run_psql(sql: str) -> str:
    env = os.environ.copy()
    env["PATH"] = f"{PSQL_PATH}:{env.get('PATH', '')}"
    cmd = ["psql", DB_NAME, "-t", "-A", "-F", "\t", "-c", sql]
    proc = subprocess.run(cmd, capture_output=True, text=True, env=env)
    if proc.returncode != 0:
        raise RuntimeError(proc.stderr.strip() or "psql query failed")
    return proc.stdout


def _parse_hhmm(s: str | None) -> int | None:
    if not s:
        return None
    try:
        hh, mm = s.strip().split(":")
        return int(hh) * 60 + int(mm)
    except Exception:
        return None


def load_patterns(days_back: int = 60) -> PatternSummary:
    sql = f"""
      SELECT
        pattern_type,
        COALESCE(value, ''),
        EXTRACT(DOW FROM timestamp AT TIME ZONE 'America/New_York')::int AS dow,
        EXTRACT(HOUR FROM timestamp AT TIME ZONE 'America/New_York')::int AS hour
      FROM cortana_patterns
      WHERE timestamp >= NOW() - INTERVAL '{days_back} days'
        AND pattern_type IN ('wake', 'sleep_check');
    """
    out = _run_psql(sql)

    wake: dict[int, list[int]] = {i: [] for i in range(7)}
    sleep: dict[int, list[int]] = {i: [] for i in range(7)}
    observed: dict[int, dict[int, int]] = {i: {} for i in range(7)}

    sample_count = 0
    for line in out.splitlines():
        if not line.strip():
            continue
        ptype, value, dow_s, hr_s = line.split("\t")
        dow = int(dow_s)
        hour = int(hr_s)
        observed[dow][hour] = observed[dow].get(hour, 0) + 1

        minutes = _parse_hhmm(value)
        if ptype == "wake" and minutes is not None:
            wake[dow].append(minutes)
            sample_count += 1
        elif ptype == "sleep_check" and minutes is not None:
            sleep[dow].append(minutes)
            sample_count += 1

    return PatternSummary(
        wake_minutes_by_dow=wake,
        sleep_minutes_by_dow=sleep,
        observed_hours_by_dow=observed,
        sample_count=sample_count,
    )


def _median_or_default(values: list[int], default_min: int) -> int:
    if not values:
        return default_min
    return int(statistics.median(values))


def _window_for_dow(summary: PatternSummary, dow: int) -> tuple[int, int]:
    # Default priors based on known routine if data is sparse.
    default_wake = 4 * 60 + 40  # 04:40
    default_sleep = 22 * 60      # 22:00

    wake_today = summary.wake_minutes_by_dow.get(dow, [])
    sleep_today = summary.sleep_minutes_by_dow.get(dow, [])

    # If day-specific sparse, back off to whole-week medians.
    if not wake_today:
        wake_today = [m for vals in summary.wake_minutes_by_dow.values() for m in vals]
    if not sleep_today:
        sleep_today = [m for vals in summary.sleep_minutes_by_dow.values() for m in vals]

    wake_min = _median_or_default(wake_today, default_wake)
    sleep_min = _median_or_default(sleep_today, default_sleep)
    return wake_min, sleep_min


def _minutes_of_day(dt: datetime) -> int:
    return dt.hour * 60 + dt.minute


def infer_state(summary: PatternSummary, now_et: datetime, urgency: str) -> dict:
    dow = now_et.weekday()
    # Python Monday=0; postgres DOW uses Sunday=0.
    pg_dow = (dow + 1) % 7

    wake_min, sleep_min = _window_for_dow(summary, pg_dow)
    minute = _minutes_of_day(now_et)

    # Activity histogram cue for likely "busy" blocks.
    hourly = summary.observed_hours_by_dow.get(pg_dow, {})
    max_seen = max(hourly.values()) if hourly else 0
    hour_density = (hourly.get(now_et.hour, 0) / max_seen) if max_seen else 0.0

    if minute < wake_min - 20 or minute >= min(24 * 60, sleep_min + 30):
        state = "sleeping"
        base_conf = 0.86
    elif wake_min - 20 <= minute < wake_min + 90:
        state = "awake"
        base_conf = 0.74
    elif sleep_min - 90 <= minute < sleep_min + 30:
        state = "winding-down"
        base_conf = 0.8
    else:
        # Workday bias: likely busy during core hours.
        if 9 <= now_et.hour <= 17 and now_et.weekday() < 5:
            state = "busy" if hour_density >= 0.25 else "available"
            base_conf = 0.66 if state == "busy" else 0.58
        else:
            state = "available"
            base_conf = 0.64

    # Confidence scales with data volume.
    data_factor = min(1.0, 0.45 + summary.sample_count / 35.0)
    confidence = round(max(0.35, min(0.97, base_conf * data_factor)), 2)

    tone = recommend_tone(state, now_et)
    alert_ok = recommend_alert_ok(state, urgency)
    next_window = next_available_window(now_et, state, wake_min, sleep_min)

    return {
        "state": state,
        "confidence": confidence,
        "recommended_tone": tone,
        "alert_ok": alert_ok,
        "next_available_window": next_window,
    }


def recommend_tone(state: str, now_et: datetime) -> str:
    h = now_et.hour
    # Required calibration anchors:
    # morning=energetic, late-night=minimal, post-workout=brief
    if 4 <= h < 8:
        return "energetic"
    if h >= 22 or h < 5:
        return "minimal"
    if 6 <= h <= 8:
        return "brief"  # likely post-workout / morning transition
    if state == "busy":
        return "brief"
    if state == "winding-down":
        return "minimal"
    return "balanced"


def recommend_alert_ok(state: str, urgency: str) -> bool:
    urgency = urgency.lower().strip()
    is_urgent = urgency in {"urgent", "critical", "high"}

    if state in {"available", "awake"}:
        return True
    if state == "busy":
        return is_urgent
    if state == "winding-down":
        return is_urgent
    if state == "sleeping":
        return urgency in {"critical"}
    return False


def _build_window(start_dt: datetime, minutes: int) -> str:
    end_dt = start_dt + timedelta(minutes=minutes)
    return f"{start_dt.isoformat()}/{end_dt.isoformat()}"


def next_available_window(now_et: datetime, state: str, wake_min: int, sleep_min: int) -> str:
    today = now_et.date()

    def at_min(day, mins):
        return datetime.combine(day, time(mins // 60, mins % 60), tzinfo=ET)

    if state in {"available", "awake"}:
        return _build_window(now_et, 90)

    if state == "busy":
        # Nudge to the top of next hour for a likely context switch.
        next_hour = (now_et.replace(minute=0, second=0, microsecond=0) + timedelta(hours=1))
        return _build_window(next_hour, 60)

    if state == "winding-down":
        tomorrow = today + timedelta(days=1)
        start = at_min(tomorrow, wake_min + 30)
        return _build_window(start, 90)

    # sleeping
    if _minutes_of_day(now_et) < wake_min:
        start = at_min(today, wake_min + 20)
    else:
        tomorrow = today + timedelta(days=1)
        start = at_min(tomorrow, wake_min + 20)
    return _build_window(start, 90)


def main() -> None:
    parser = argparse.ArgumentParser(description="Predict Chief state and messaging timing/tone")
    parser.add_argument("--urgency", default="normal", choices=["low", "normal", "high", "urgent", "critical"])
    parser.add_argument("--at", help="ISO datetime in ET or with offset; default now")
    parser.add_argument("--days-back", type=int, default=60)
    args = parser.parse_args()

    if args.at:
        dt = datetime.fromisoformat(args.at)
        now_et = dt.astimezone(ET) if dt.tzinfo else dt.replace(tzinfo=ET)
    else:
        now_et = datetime.now(ET)

    summary = load_patterns(days_back=args.days_back)
    result = infer_state(summary, now_et, args.urgency)

    print(json.dumps(result, separators=(",", ":")))


if __name__ == "__main__":
    main()

File: tools/behavioral-twin/predict.ts
#!/usr/bin/env npx tsx

import { spawnSync } from "node:child_process";
import { withPostgresPath } from "../lib/db.js";

type PatternSummary = {
  wakeMinutesByDow: Record<number, number[]>;
  sleepMinutesByDow: Record<number, number[]>;
  observedHoursByDow: Record<number, Record<number, number>>;
  sampleCount: number;
};

const DB_NAME = process.env.CORTANA_DB ?? "cortana";

function runPsql(sql: string): string {
  const proc = spawnSync("psql", [DB_NAME, "-t", "-A", "-F", "\t", "-c", sql], {
    encoding: "utf8",
    env: withPostgresPath(process.env),
  });
  if (proc.status !== 0) throw new Error((proc.stderr || "psql query failed").trim());
  return proc.stdout;
}

function parseHHMM(s?: string): number | null {
  if (!s) return null;
  const m = s.trim().match(/^(\d{1,2}):(\d{2})$/);
  if (!m) return null;
  return Number(m[1]) * 60 + Number(m[2]);
}

function loadPatterns(daysBack = 60): PatternSummary {
  const sql = `
      SELECT pattern_type, COALESCE(value, ''),
      EXTRACT(DOW FROM timestamp AT TIME ZONE 'America/New_York')::int AS dow,
      EXTRACT(HOUR FROM timestamp AT TIME ZONE 'America/New_York')::int AS hour
      FROM cortana_patterns
      WHERE timestamp >= NOW() - INTERVAL '${daysBack} days'
      AND pattern_type IN ('wake', 'sleep_check');
  `;
  const out = runPsql(sql);
  const wake: Record<number, number[]> = { 0: [], 1: [], 2: [], 3: [], 4: [], 5: [], 6: [] };
  const sleep: Record<number, number[]> = { 0: [], 1: [], 2: [], 3: [], 4: [], 5: [], 6: [] };
  const observed: Record<number, Record<number, number>> = { 0: {}, 1: {}, 2: {}, 3: {}, 4: {}, 5: {}, 6: {} };

  let sampleCount = 0;
  for (const line of out.split("\n")) {
    if (!line.trim()) continue;
    const [ptype, value, dowS, hrS] = line.split("\t");
    const dow = Number(dowS);
    const hour = Number(hrS);
    observed[dow][hour] = (observed[dow][hour] ?? 0) + 1;

    const minutes = parseHHMM(value);
    if (ptype === "wake" && minutes !== null) {
      wake[dow].push(minutes);
      sampleCount += 1;
    } else if (ptype === "sleep_check" && minutes !== null) {
      sleep[dow].push(minutes);
      sampleCount += 1;
    }
  }

  return { wakeMinutesByDow: wake, sleepMinutesByDow: sleep, observedHoursByDow: observed, sampleCount };
}

const medianOrDefault = (values: number[], def: number): number => {
  if (values.length === 0) return def;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 ? sorted[mid] : Math.floor((sorted[mid - 1] + sorted[mid]) / 2);
};

function windowForDow(summary: PatternSummary, dow: number): [number, number] {
  const defaultWake = 4 * 60 + 40;
  const defaultSleep = 22 * 60;

  let wake = summary.wakeMinutesByDow[dow] ?? [];
  let sleep = summary.sleepMinutesByDow[dow] ?? [];

  if (!wake.length) wake = Object.values(summary.wakeMinutesByDow).flat();
  if (!sleep.length) sleep = Object.values(summary.sleepMinutesByDow).flat();

  return [medianOrDefault(wake, defaultWake), medianOrDefault(sleep, defaultSleep)];
}

const minuteOfDay = (d: Date) => d.getHours() * 60 + d.getMinutes();

function recommendTone(state: string, now: Date): string {
  const h = now.getHours();
  if (h >= 4 && h < 8) return "energetic";
  if (h >= 22 || h < 5) return "minimal";
  if (h >= 6 && h <= 8) return "brief";
  if (state === "busy") return "brief";
  if (state === "winding-down") return "minimal";
  return "balanced";
}

function recommendAlertOk(state: string, urgency: string): boolean {
  const u = urgency.trim().toLowerCase();
  const urgent = ["urgent", "critical", "high"].includes(u);
  if (["available", "awake"].includes(state)) return true;
  if (state === "busy" || state === "winding-down") return urgent;
  if (state === "sleeping") return u === "critical";
  return false;
}

function buildWindow(start: Date, minutes: number): string {
  const end = new Date(start.getTime() + minutes * 60000);
  return `${start.toISOString()}/${end.toISOString()}`;
}

function nextAvailableWindow(now: Date, state: string, wakeMin: number): string {
  const mk = (base: Date, mins: number): Date => {
    const d = new Date(base);
    d.setHours(Math.floor(mins / 60), mins % 60, 0, 0);
    return d;
  };

  if (state === "available" || state === "awake") return buildWindow(now, 90);
  if (state === "busy") {
    const n = new Date(now);
    n.setMinutes(0, 0, 0);
    n.setHours(n.getHours() + 1);
    return buildWindow(n, 60);
  }
  if (state === "winding-down") {
    const t = new Date(now);
    t.setDate(t.getDate() + 1);
    const start = mk(t, wakeMin + 30);
    return buildWindow(start, 90);
  }

  const m = minuteOfDay(now);
  if (m < wakeMin) return buildWindow(mk(now, wakeMin + 20), 90);
  const t = new Date(now);
  t.setDate(t.getDate() + 1);
  return buildWindow(mk(t, wakeMin + 20), 90);
}

function inferState(summary: PatternSummary, now: Date, urgency: string) {
  const jsDow = now.getDay();
  const [wakeMin, sleepMin] = windowForDow(summary, jsDow);
  const minute = minuteOfDay(now);

  const hourly = summary.observedHoursByDow[jsDow] ?? {};
  const maxSeen = Math.max(0, ...Object.values(hourly));
  const hourDensity = maxSeen ? (hourly[now.getHours()] ?? 0) / maxSeen : 0;

  let state = "available";
  let baseConf = 0.64;

  if (minute < wakeMin - 20 || minute >= Math.min(24 * 60, sleepMin + 30)) {
    state = "sleeping";
    baseConf = 0.86;
  } else if (minute >= wakeMin - 20 && minute < wakeMin + 90) {
    state = "awake";
    baseConf = 0.74;
  } else if (minute >= sleepMin - 90 && minute < sleepMin + 30) {
    state = "winding-down";
    baseConf = 0.8;
  } else if (now.getHours() >= 9 && now.getHours() <= 17 && now.getDay() >= 1 && now.getDay() <= 5) {
    state = hourDensity >= 0.25 ? "busy" : "available";
    baseConf = state === "busy" ? 0.66 : 0.58;
  }

  const dataFactor = Math.min(1, 0.45 + summary.sampleCount / 35);
  const confidence = Math.round(Math.max(0.35, Math.min(0.97, baseConf * dataFactor)) * 100) / 100;

  return {
    state,
    confidence,
    recommended_tone: recommendTone(state, now),
    alert_ok: recommendAlertOk(state, urgency),
    next_available_window: nextAvailableWindow(now, state, wakeMin),
  };
}

function parseArgs(argv: string[]) {
  const out = { urgency: "normal", at: "", daysBack: 60 };
  for (let i = 0; i < argv.length; i += 1) {
    const a = argv[i];
    if (a === "--urgency" && argv[i + 1]) out.urgency = argv[++i];
    else if (a === "--at" && argv[i + 1]) out.at = argv[++i];
    else if (a === "--days-back" && argv[i + 1]) out.daysBack = Number(argv[++i]);
  }
  return out;
}

function main(): void {
  const args = parseArgs(process.argv.slice(2));
  const allowed = new Set(["low", "normal", "high", "urgent", "critical"]);
  if (!allowed.has(args.urgency)) {
    process.stderr.write("Invalid --urgency\n");
    process.exit(2);
  }

  const now = args.at ? new Date(args.at) : new Date();
  if (Number.isNaN(now.getTime())) {
    process.stderr.write("Invalid --at datetime\n");
    process.exit(2);
  }

  const summary = loadPatterns(args.daysBack);
  const result = inferState(summary, now, args.urgency);
  process.stdout.write(`${JSON.stringify(result)}\n`);
}

main();
